design-documents/atomics-ABI.rst

```diff
@@ -0,0 +1,302 @@
+Program we see there is one additional outcome of executing the compiled
+program that is not an outcome of executing the Concurrent Program. This is 
+because the Arm Architecture Reference Manual [ARMARM_] states that 
+*instructions where the destination register is WZR or XZR, are not regarded 
+as doing a read for the purpose of a DMB LD barrier.*
+
+In this case the compiler introduces another outcome of Execution. To fix this
+issue, a compiler is not permitted to rewrite the destination register to be the
+zero register::
+
+  thread_0:
+    MOV W9,#1
+    STR W9,[X2]
+    DMB ISH
+    STR W3,[X4]
+
+  thread_1:
+    MOV W9,#2
+    SWP W9, W10, [X2]
+    DMB ISHLD
+    LDR W3,[X4]
+
+Executing the compiled program on an Arm-based machine from a fixed initial
+state (where ``x`` and ``y`` are ``0``) produces one of the following outcomes,
+according to the AArch64 Memory Model contained in §B2 of the Arm Architecture
+Reference Manual [ARMARM_]::
+
+  { thread_1:r0=0; [y]=1; }
+  { thread_1:r0=1; [y]=1; }
+  { thread_1:r0=1; [y]=2; }
+
+As such the unexpected outcome has disappeared. There are multiple Mappings
+that exhibit this behaviour. Assembly Sequences affected make use of ``SWP`` 
+and ``LD<OP>`` Assembly instructions.
+
+Documentation
+~~~~~~~~~~~~~
+
+The collective knowledge of atomics ABIs exists as numerous online discusions.
+These discussions are neither authoritative nor persistent. Some discussions 
+are now inaccessible and others are out of date. This is problematic given the
+inherent complexity of relaxed memory concurrency, the difficulty of finding bugs,
+and the possibility of user error. We believe an ABI is necessary to document
+this corner of code generation.
+
+
+The Mix Testing Process
+-----------------------
+
+ABI compatibility must be testable. Concurrency is not trivial, and the ABI
+presents a simplification of part of the problem that is understandable by
+engineers. We provide a simple technique for testing ABI compatibility.
+These techniques reduce the difficulty of checking compatibility from a 
+problem of understanding concurrent executions, to the familiar testing 
+domain of comparing program outcomes of tests. This document does not 
+preclude other means of testing compatibility.
+
+We test for Compiler bugs. A Compiler Bug is defined as an outcome of a
+compiled program execution (under the AArch64 Memory Model contained in
+§B2 of the Arm Architecture Reference Manual [ARMARM_]) that is not 
+an outcome of execution of the source Concurrent Program (under the 
+ISO C memory model). Consider the hypothetical example where a source
+Concurrent Program finishes execution in one of three possible outcomes
+(a reference for this notation is found here [PAPER_])::
+
+  { thread_0:r0=0, thread_1:r0=1 }
+  { thread_0:r0=1, thread_1:r0=0 }
+  { thread_0:r0=1, thread_1:r0=1 }
+
+and one possible compiled program outcome has the following outcomes 
```

Comment:
`one possible compiled program outcome has the following outcomes`. double outcomes.

Could this be rephrased to: `one compiled program execution run has the following possible outcomes`